tuf/signed/sign.go

```diff
@@ -13,73 +13,76 @@ package signed
 
 import (
 	"crypto/rand"
-	"fmt"
 
 	"github.com/Sirupsen/logrus"
+	"github.com/docker/notary/trustmanager"
 	"github.com/docker/notary/tuf/data"
 	"github.com/docker/notary/tuf/utils"
 )
 
-// Sign takes a data.Signed and a key, calculated and adds the signature
-// to the data.Signed
-func Sign(service CryptoService, s *data.Signed, keys ...data.PublicKey) error {
-	logrus.Debugf("sign called with %d keys", len(keys))
-	signatures := make([]data.Signature, 0, len(s.Signatures)+1)
-	signingKeyIDs := make(map[string]struct{})
-	ids := make([]string, 0, len(keys))
+// Sign takes a data.Signed and replaces data.Signature with at least
+// minPrimarySignature signatures using primaryKeys and any possible
+// signatures using optionalKeys
+func Sign(service CryptoService, s *data.Signed, primaryKeys []data.PublicKey,
+	minSignatures int, optionalKeys []data.PublicKey) error {
+	logrus.Debugf("sign called with %d/%d required + %d optional keys",
+		minSignatures, len(primaryKeys), len(optionalKeys))
 
 	privKeys := make(map[string]data.PrivateKey)
 
 	// Get all the private key objects related to the public keys
-	for _, key := range keys {
+	missingKeyIDs := []string{}
+	for _, key := range primaryKeys {
 		canonicalID, err := utils.CanonicalKeyID(key)
-		ids = append(ids, canonicalID)
 		if err != nil {
-			continue
+			return err
 		}
 		k, _, err := service.GetPrivateKey(canonicalID)
 		if err != nil {
-			continue
+			if _, ok := err.(trustmanager.ErrKeyNotFound); ok {
```

Comment:
> Absence of a key (for any reason) is not sufficient reason to fail signing entirely.
_Absence_ of a key, no.

For a corrupted key or key store, doesn’t it make more sense to fail early (e.g. while recent backups are still available)?

We could have `ErrInsufficientSignatures` collect a map of key ID → failure reason, but that’s a fairly non-obvious error reporting model: Are UIs prepared to deal with a kilobyte of text if 5 keys fail? Can we even make this readable without newlines? And all of this for a hopefully very rare storage or HSM failure.